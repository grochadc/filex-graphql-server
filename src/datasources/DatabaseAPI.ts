import { ApolloError } from "apollo-server";
import { DataSource } from "apollo-datasource";
import { Workshop, Teacher, TeacherOption } from "../generated/graphql";
import { ParameterizedQuery as PQ } from "pg-promise";

export interface OptionModel {
  id: number;
  day: string;
  time: string;
  teacher_name: string;
  url: string;
  workshop_name: string;
  workshop_description: string;
  workshop_levels: number[];
  teacher_id: number;
  workshop_id: number;
  reservations: number;
}

export type StudentModel = {
  id: number;
  codigo: string;
  nombre: string;
  apellido_paterno: string;
  apellido_materno: string;
  genero: string;
  carrera: string;
  ciclo: string;
  telefono: string;
  email: string;
  nivel: string;
  curso: string;
  externo: boolean;
  grupo: string;
};

type Option = Omit<
  OptionModel,
  "id" | "workshop_id" | "teacher_id" | "reservations"
> & {
  id: string;
  workshop_id: string;
  teacher_id: string;
  isTutorial: boolean;
  available: boolean;
};

export const SELECT_STUDENT = `SELECT * FROM student WHERE codigo=$1`;

export const SELECT_STUDENT_RESERVATION = `
SELECT
  r.*,
  o.day,
  o.time,
  o.url,
  o.teacher_id,
  t.name as teacher_name,
  o.workshop_id,
  w.name as workshop_name
FROM reservation r
INNER JOIN option o ON o.id=r.option_id
INNER JOIN teacher t ON t.id=o.teacher_id
INNER JOIN workshop w ON w.id=o.workshop_id
WHERE r.student_id=$1
`;

export const SELECT_NEW_RESERVATION = `
SELECT
  r.id,
  o.day,
  o.time,
  t.name as teacher_name,
  t.id as teacher_id,
  w.name as workshop_name,
  w.id as workshop_id,
  o.url
FROM reservation r
INNER JOIN option o ON r.option_id=o.id
INNER JOIN teacher t ON o.teacher_id=t.id
INNER JOIN workshop w ON o.workshop_id=w.id
WHERE r.id=$1
`;

export const SELECT_OPTIONS = `
SELECT
  Option.id,
  Workshop.name as workshop_name,
  Option.day,
  Option.time,
  Teacher.name as teacher_name,
  Option.url,
  Workshop.description as workshop_description,
  Workshop.levels as workshop_levels,
  Teacher.id as teacher_id,
  Workshop.id as workshop_id,
  COUNT(Reservation.id) as reservations
FROM Option
INNER JOIN Teacher ON Option.teacher_id=Teacher.id
INNER JOIN Workshop on Option.workshop_id=Workshop.id
LEFT JOIN Reservation ON Option.id=Reservation.option_id
GROUP BY Option.id, Workshop.id, Teacher.id
ORDER BY Option.id;
`;

export const SELECT_TEACHER_INFO = `
SELECT
  id,
  name,
  json_agg(
        (
        SELECT
          json_build_object('option_id',r.option_id, 'students', json_agg(s))
        FROM reservation r
        INNER JOIN student s ON s.id=r.student_id
        WHERE r.option_id=any(teacher.option_ids)
        GROUP BY r.option_id
        )
  ) as options FROM teacher Where id=$1 GROUP BY teacher.id;
`;

export type TeacherReservation = {
  reservation_id: number;
  codigo: string;
  nombre: string;
  apellido_materno: string;
  apellido_paterno: string;
  teacher_id: number;
  teacher: string;
  workshop: string;
  time: string;
  day: string;
  option_id: string;
};
export const SELECT_TEACHER_RESERVATIONS = `
SELECT
  t.id as teacher_id,
  r.id as reservation_id,
  codigo,
  nombre,
  apellido_paterno,
  apellido_materno,
  t.id as teacher_id,
  t.name as teacher,
  w.name as workshop,
  o.time,
  o.day,
  o.url,
  o.id as option_id
FROM teacher t
INNER JOIN option o ON o.teacher_id=t.id
INNER JOIN workshop w ON o.workshop_id=w.id
INNER JOIN reservation r ON r.option_id=o.id
INNER JOIN student s ON s.id=r.student_id
WHERE t.id=$1
`;

export const SELECT_SINGLE_OPTION = `
SELECT * FROM Option o WHERE o.id=$1;
`;

class DatabaseAPI extends DataSource {
  db: any;
  constructor(db: any) {
    super();
    console.log("construting databaseAPI");
    this.db = db;
  }
  _getStudentSql(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    return this.db.any(paramQuery);
  }
  async getStudent(codigo: string) {
    if (/^[1234567890]+$/.test(codigo)) {
      const result: StudentModel[] = await this._getStudentSql(SELECT_STUDENT, [
        codigo
      ]);
      if (result.length === 0)
        throw new ApolloError("404: Alumno inexistente", "404");
      return result[0];
    } else {
      throw new ApolloError("Input wasn't a number");
    }
  }
  async getOptionsByIDs(ids: string[]): Promise<Option> {
    const parsedIDs = ids.map(id => Number(id));
    const result = await this.db.any(
      new PQ({ text: SELECT_SINGLE_OPTION, values: [parsedIDs] })
    );
    return result[0];
  }
  async getAllOptions(max_reservations: number): Promise<Option[]> {
    const options: OptionModel[] = await this.db.any(SELECT_OPTIONS);
    const mapped = options.map(option => {
      return {
        ...option,
        available: Boolean(Number(option.reservations) < max_reservations),
        isTutorial: Boolean(option.workshop_id === 4),
        id: option.id.toString(),
        workshop_id: option.workshop_id.toString(),
        teacher_id: option.teacher_id.toString()
      };
    });
    return mapped;
  }

  async getAllWorkshops(max_reservations: number): Promise<Workshop[]> {
    const options = await this.getAllOptions(max_reservations);
    const workshops: Workshop[] = [];
    options.forEach(option => {
      const { workshop_description, ...simpleOption } = option;
      const foundWorkshopIndex = workshops.findIndex(
        workshop => workshop.id === option.workshop_id
      );
      if (foundWorkshopIndex === -1) {
        const new_workshop = {
          id: option.workshop_id,
          name: option.workshop_name,
          description: option.workshop_description,
          levels: option.workshop_levels.map(level => level.toString()),
          options: [simpleOption]
        };
        workshops.push(new_workshop);
      } else {
        workshops[foundWorkshopIndex].options.push(simpleOption);
      }
    });
    return workshops;
  }
  _getStudentReservationSql(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    return this.db.any(paramQuery);
  }
  getStudentReservation(id: number) {
    return this._getStudentReservationSql(SELECT_STUDENT_RESERVATION, [id]);
  }
  async _makeReservation(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    const result = await this.db.any(paramQuery);
    return result;
  }
  _newReservation(id: number) {
    return this.db.any(new PQ({ text: SELECT_NEW_RESERVATION, values: [id] }));
  }
  async makeReservation(
    student_id: number,
    option_id: number,
    tutorial_reason?: string
  ) {
    const INSERT_RESERVATION = `
      INSERT INTO reservation (student_id, option_id, tutorial_reason) VALUES ($1,$2,$3) RETURNING id;
    `;
    const normalizedTutorialReason =
      tutorial_reason === undefined ? null : tutorial_reason;
    const reservations = await this._makeReservation(INSERT_RESERVATION, [
      student_id,
      option_id,
      normalizedTutorialReason
    ]);
    const result = await this._newReservation(reservations[0].id);
    return result;
  }
  _resetReservations(query) {
    return this.db.any(query);
  }
  async resetReservations() {
    const DELETE_RESERVATIONS = `DELETE FROM reservation`;
    try {
      this._resetReservations(DELETE_RESERVATIONS);
    } catch (e) {
      throw new ApolloError(e);
    }
    return true;
  }
  _getTeacher(id: number) {
    const parsedQuery = new PQ({
      text: SIMPLE_TEACHER_INFO,
      values: [id]
    });
    return this.db.any(parsedQuery);
  }
  async getTeacher(id: string) {
    const parsedID = Number(id);
    const result: SimpleTeacherInfo[] = await this._getTeacher(parsedID);
    return {
      id: result[0].id.toString(),
      name: result[0].name,
      option_ids: result[0].option_ids.map(id => id.toString()),
      options: result.map(teacher => {
        return {
          id: teacher.option_id.toString(),
          day: teacher.day,
          time: teacher.time,
          teacher_name: teacher.name,
          teacher_id: teacher.id.toString(),
          workshop_name: teacher.workshop_name,
          workshop_id: teacher.workshop_id.toString(),
          url: teacher.url
        };
      })
    };
  }
  async getAllTeachers() {
    const result: SelectTeachers = await this.db.any(SELECT_TEACHERS);
    return result.map(teacher => {
      return {
        ...teacher,
        id: teacher.id.toString(),
        option_ids: teacher.option_ids.map(id => id.toString())
      };
    });
  }
  _getTeacherReservations(option_id: number) {
    return this.db.any(
      new PQ({ text: GET_TEACHER_RESERVATIONS, values: [option_id] })
    );
  }
  getTeacherReservations(option_id: string) {
    const parsedOptionID = Number(option_id);
    return this._getTeacherReservations(parsedOptionID);
  }
  _deleteOptionReservations(option_id: number) {
    return this.db.any(
      new PQ({ text: DELETE_TEACHER_RESERVATIONS, values: [option_id] })
    );
  }
  deleteOptionReservations(option_id: string) {
    const parsedID = Number(option_id);
    this._deleteOptionReservations(parsedID);
  }
}

export const GET_TEACHER_RESERVATIONS = `
SELECT
  r.id,
  w.id as workshop_id,
  w.name as workshop_name,
  o.id as option_id,
  s.codigo,
  s.nombre,
  s.apellido_paterno,
  s.apellido_materno,
  s.nivel,
  s.grupo,
  r.tutorial_reason
FROM reservation r
INNER JOIN option o ON r.option_id=o.id
INNER JOIN workshop w ON o.workshop_id=w.id
INNER JOIN student s ON s.id=r.student_id
WHERE o.id=$1
`;
export type SelectTeachers = {
  id: number;
  name: string;
  option_ids: number[];
}[];
export const SELECT_TEACHERS = `
SELECT * FROM Teacher;
`;

export type SimpleTeacherInfo = {
  id: number;
  name: string;
  workshop_name: string;
  workshop_id: number;
  option_id: number;
  day: string;
  time: string;
  url: string;
  option_ids: number[];
};
export const SIMPLE_TEACHER_INFO = `
SELECT t.id, t.name, w.name as workshop_name, w.id as workshop_id, o.id as option_id, o.day, o.time, o.url, t.option_ids
FROM teacher t
INNER JOIN option o ON o.teacher_id=t.id
INNER JOIN workshop w ON w.id=o.workshop_id
WHERE t.id=$1
`;

export const DELETE_TEACHER_RESERVATIONS = `
  DELETE FROM reservation WHERE option_id=$1
`;

export default DatabaseAPI;
