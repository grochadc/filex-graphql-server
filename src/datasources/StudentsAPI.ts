import { RESTDataSource } from "apollo-datasource-rest";

export type GetStudentFn = (codigo: string, env?: 'dev'|'prod') => Promise<any>;

class StudentsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/students";
  }

  async _get(url: string, env:string|undefined) {
    const finalUrl = env ? '/'+env+url : '/prod'+url;
    return this.get(finalUrl);
  }

  async getStudent(...[codigo, env]: Parameters<GetStudentFn>) {
    return this.get(`/${env}/${codigo}.json`);
  }
}

export { StudentsAPI };
