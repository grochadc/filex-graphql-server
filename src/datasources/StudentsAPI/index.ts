import { DataSource } from "apollo-datasource";
import { StudentModel } from "../../modules/workshops/models";
import { ParameterizedQuery as PQ } from "pg-promise";
import { ApolloError } from "apollo-server";
import { PrismaClient } from "@prisma/client";
import {
  StudentInput,
  ApplicantInput,
  Applicant,
} from "../../generated/graphql";

type MyStudentInput = Omit<StudentModel, "id">;
type Changes = Partial<MyStudentInput>;

class StudentsAPI extends DataSource {
  db: any;
  prisma: PrismaClient;
  constructor(db: any, prisma: PrismaClient) {
    super();
    if (prisma === undefined)
      throw new Error("Include a new PrismaCLient() on constructor");
    this.prisma = prisma;
    this.db = db;
    if (prisma === undefined) {
      throw Error("Supply a new PrismaCLient() on constructor");
    }
    this.prisma = prisma;
  }

  async getAllStudents(cicloActual: string) {
    const students = await this.prisma.student.findMany({
      where: {
        cicloActual: cicloActual,
        groupId: {
          not: null
        }
      },
      include: {
        applicant: true,
        groupObject: {
          include: {
            teacher: true,
          },
        },
      },
    });

    return students;
  }

  async getStudent(codigo: string, cicloActual: string) {
    if (codigo === null)
      throw new Error("Codigo not provided on studentsAPI.getStudent");
    if (cicloActual === null)
      throw new Error("ciclo_actual not provided on studentsAPI.getStudent");
    const student = await this.prisma.student.findFirst({
      include: {
        applicant: true,
        groupObject: true,
      },
      where: {
        codigo,
        cicloActual,
      },
    });
    if (student === null) throw new Error("Student not found on database");
    if (student.applicant == null)
      throw new Error("Applicant info is missing on database");
    if (student.groupObject == null)
      throw new Error("Couldnt find a group Object");

    return student;
  }

  async addApplicant(
    applicant: ApplicantInput
  ): Promise<Omit<Applicant, "curso" | "groups" | "nivel" | "registering">> {
    const result = await this.prisma.applicant.create({
      data: applicant,
    });

    return {
      ...result,
      id: String(result.id),
    };
  }

  async addStudent(student: StudentInput) {
    const applicant = await this.prisma.applicant.findFirst({
      where: {
        codigo: student.codigo,
      },
    });

    if (applicant === null)
      throw new Error(
        "No applicant found on database. Please submit the student's personal information on the Applicant endpoint first."
      );

    const groupObj = await this.prisma.group.findFirst({
      where: {
        name: student.grupo,
        ciclo: student.cicloActual,
      },
    });
    return Promise.resolve();
    /*
    return this.prisma.student.create({
      data: {
        applicant: { connect: {id: applicant.id} },
        codigo: student.codigo,
        curso: student.curso,
        groupObject: { connect: { id: groupObj.id } },
        ciclo_actual: student.cicloActual,
        nivel: student.nivel
      }
    })
    */
  }

  async editStudent(codigo: string, changes: Changes, ciclo: string) {
    const oldStudent = await this.getStudent(codigo, ciclo);
    const newStudent = Object.assign(oldStudent, changes);
    try {
      this.db.none(
        new PQ({
          text: EDIT_STUDENT,
          values: [
            newStudent.codigo,
            newStudent.nombre,
            newStudent.apellido_materno,
            newStudent.apellido_paterno,
            newStudent.genero,
            newStudent.carrera,
            newStudent.ciclo,
            newStudent.telefono,
            newStudent.email,
            newStudent.nivel,
            newStudent.grupo,
            newStudent.externo,
            newStudent.curso,
            newStudent.id,
          ],
        })
      );
    } catch (e) {
      console.error(e);
    }
    return newStudent;
  }
}

export const GET_STUDENT = `
  SELECT * FROM Student WHERE codigo=$1
`;

export const ADD_STUDENT = `
  INSERT INTO Student (
    codigo,
    nombre,
    apellido_materno,
    apellido_paterno,
    genero,
    carrera,
    ciclo,
    telefono,
    email,
    nivel,
    grupo,
    externo,
    curso
  ) values (
    $1,
    $2,
    $3,
    $4,
    $5,
    $6,
    $7,
    $8,
    $9,
    $10,
    $11,
    $12,
    $13
  )
  RETURNING id
`;

export const EDIT_STUDENT = `
  UPDATE Student SET
  codigo=$1,
  nombre= $2,
  apellido_materno= $3,
  apellido_paterno= $4,
  genero= $5,
  carrera= $6,
  ciclo= $7,
  telefono= $8,
  email= $9,
  nivel= $10,
  grupo= $11,
  externo= $12,
  curso= $13
  WHERE id=$14
`;

export { StudentsAPI };
