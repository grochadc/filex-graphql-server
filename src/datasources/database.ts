import pgPromise from "pg-promise";
import path from "path";
const { QueryFile } = pgPromise;
require("dotenv").config();

const initOptions = {
  connectionString: process.env.DATABASE_URL,
  ssl: { rejectUnauthorized: false }
};

console.log("Initializing database");
const pgp = pgPromise();
const db = pgp(initOptions);

export function sqlFile(file: string) {
  const fullPath = path.join(__dirname, file);
  const options = {
    minify: true
  };
  const qf = new QueryFile(fullPath, options);

  if (qf.error) {
    console.error(qf.error);
  }

  return qf;
}

export default db;
