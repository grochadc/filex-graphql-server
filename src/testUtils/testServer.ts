import {
  createTestClient,
  ApolloServerTestClient
} from "apollo-server-testing";
import { ApolloServer } from "apollo-server";

export default function testServer(
  dataSources?: () => any,
  context?: () => any
): ApolloServerTestClient {
  return createTestClient(
    new ApolloServer({
      modules: [
        require("../modules/placement_exam"),
        require("../modules/placement_settings"),
        require("../modules/registro"),
        require("../modules/workshops"),
        require("../modules/test_questions")
      ],
      dataSources,
      context
    })
  );
}
