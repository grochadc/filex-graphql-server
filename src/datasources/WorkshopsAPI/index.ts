import { RESTDataSource } from "apollo-datasource-rest";
import { AttendingStudent } from "../../generated/graphql";
import {
  WorkshopModel,
  OptionModel,
  DatabaseModel,
} from "../../modules/workshops/models";
import { PrismaClient } from "@prisma/client";
import { ApolloError } from "apollo-server";
import { students } from "datasources/StudentsAPI/mocks";

type Maybe<T> = T | null;

class WorkshopsAPI extends RESTDataSource {
  prisma: PrismaClient;
  constructor(prisma: PrismaClient) {
    super();
    if (prisma === undefined)
      throw new Error("Must include a new PrismaClient() on constrctor");
    this.prisma = prisma;
    this.baseURL = "https://filex-5726c.firebaseio.com/workshops/";
    if (process.env.NODE_ENV === "test") this.initialize({} as any);
  }

  getTodaysDate(): Date {
    return new Date();
  }

  async isOpen(): Promise<boolean> {
    const open = await this.get(
      `${this.context.enviroment}/system/isOpen/result.json`
    );
    return open ? open : false;
  }

  async toggleOpen(): Promise<boolean> {
    const open = await this.isOpen();
    await this.put(`${this.context.enviroment}/system/isOpen.json`, {
      result: !open,
    });
    return !open;
  }

  getTeacher(id: string) {
    return this.prisma.teacher.findUnique({
      where: {
        id: Number(id),
      },
    });
  }

  async getAllTeachers() {
    return this.prisma.teacher.findMany({
      include: {
        options: {
          include: {
            workshop: true,
          },
        },
      },
    });
  }

  async getReservationsByOptionId(option_id: string) {
    const openDate = await this.getOpenDate();
    const reservations = await this.prisma.workshopReservation.findMany({
      where: {
        option_id: Number(option_id),
        create_time: {
          gte: new Date(openDate),
        },
      },
      include: {
        student: {
          include: {
            applicant: true,
          },
        },
        option: true,
      },
    });

    if (null) return [];

    return reservations.map((reservation) => {
      return {
        ...reservation,
        id: String(reservation.id),
      };
    });
  }

  async getReservationCount(student_id: string) {
    return this.prisma.workshopReservation.count({
      where: {
        student_id: Number(student_id),
      },
    });
  }

  async getOptionReservationCount(option_id: string) {
    const openDate = await this.getOpenDate();

    return this.prisma.workshopReservation.count({
      where: {
        option_id: Number(option_id),
        create_time: {
          gte: new Date(openDate),
        },
      },
    });
  }

  async getAllOptions(max_students: number) {
    return this.prisma.workshopOption.findMany();
  }

  async getAllWorkshops(max_students: number) {
    return this.prisma.workshop.findMany({
      include: {
        options: {
          include: {
            teacher: true,
          },
        },
      },
    });
  }

  async getTeacherReservations(teacherId: number, optionId: number) {
    const openDateStr = await this.getOpenDate();
    return this.prisma.workshopReservation.findMany({
      where: {
        option: {
          teacher_id: teacherId,
          id: optionId,
        },
        create_time: {
          gt: new Date(openDateStr),
        },
      },
      include: {
        option: true,
        student: {
          include: {
            applicant: true,
            groupObject: true,
          },
        },
      },
    });
  }

  async getTeacherOptions(teacherId: string) {
    return this.prisma.workshopOption.findMany({
      include: {
        workshop: true,
      },
      where: {
        teacher_id: Number(teacherId),
      },
    });
  }

  async getStudentReservation(studentId: string) {
    const openDateString = await this.getOpenDate();
    const reservation = await this.prisma.workshopReservation.findFirst({
      where: {
        student_id: Number(studentId),
        create_time: {
          gte: new Date(openDateString),
        },
      },
      include: {
        student: true,
        option: {
          include: {
            workshop: true,
            teacher: true,
          },
        },
      },
    });
    if (!reservation) return null;
    return { ...reservation, id: String(reservation.id) };
  }

  async makeReservation(student_id: string, option_id: string) {
    if (!student_id)
      throw new Error(
        "No student_id provided to make a reservation on workshopsAPI.makeReservation"
      );
    if (!option_id)
      throw new Error(
        "No option_id provided to make a reservation on workshopsAPI.makeReservation"
      );
    const currentReservation = await this.getStudentReservation(student_id);
    if (currentReservation)
      throw new ApolloError("Student already has a reservation", "409");
    const reservation = await this.prisma.workshopReservation.create({
      data: {
        student_id: Number(student_id),
        option_id: Number(option_id),
      },
      include: {
        student: true,
        option: {
          include: {
            teacher: true,
            workshop: true,
          },
        },
      },
    });
    return { ...reservation, id: String(reservation.id) };
  }

  async resetReservations() {
    const now = await this.getTodaysDate();
    const nowString = now.toISOString();
    await this.put("system/openDate.json", { result: nowString });
    return true;
  }

  async getOpenDate(): Promise<string> {
    const openDate = await this.get("system/openDate.json");
    const dateString = openDate.result;
    return dateString;
  }

  async saveAttendance(attendance: AttendingStudent[]) {
    const reservation_ids = attendance
      .filter((student) => student.attended)
      .map((student) => Number(student.reservation_id));
    const modified = await this.prisma.workshopReservation.updateMany({
      where: {
        id: {
          in: reservation_ids,
        },
      },
      data: {
        attended: true,
      },
    });
    return modified.count > 0;
  }

  async getMaxStudentReservations() {
    const result = await this.get(`system.json`);
    return result.max_reservations;
  }

  async getReservationLimit(): Promise<number> {
    return this.get("/reservationLimit.json");
  }
}

export { WorkshopsAPI };
