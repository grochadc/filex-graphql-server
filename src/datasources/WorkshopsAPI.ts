import { RESTDataSource } from "apollo-datasource-rest";
import * as R from "ramda";
import * as utils from "../utils";
import database, { Option } from "./db";
import { Workshop, AttendingStudent } from "../generated/graphql";
import {
  WorkshopModel,
  TeacherModel,
  OptionModel,
  StudentReservationModel,
  TeacherReservationModel,
  DatabaseModel,
} from "../modules/workshops/models";

type Maybe<T> = T | null;

class WorkshopsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/workshops/";
  }

  async isOpen(): Promise<boolean> {
    const open = await this.get(
      `${this.context.enviroment}/system/isOpen/result.json`
    );
    return open ? open : false;
  }

  async toggleOpen(): Promise<boolean> {
    const open = await this.isOpen();
    await this.put(`${this.context.enviroment}/system/isOpen.json`, {
      result: !open,
    });
    return !open;
  }

  async getWorkshops(): Promise<WorkshopModel[]> {
    const workshops = await this.get(
      `${this.context.enviroment}/workshops.json`
    );
    if (workshops === null)
      throw new Error(
        `${this.context.enviroment}/workshops.json returned null`
      );
    return Object.values(workshops);
  }

  async getOptions(): Promise<DatabaseModel["options"]> {
    return (await this.get(
      `${this.context.enviroment}/options.json`
    )) as DatabaseModel["options"];
  }

  getAvailableOptions(): Promise<Maybe<DatabaseModel["availableOptions"]>> {
    return this.get(
      `${this.context.enviroment}/availableOptions.json`
    ) as Promise<Maybe<DatabaseModel["availableOptions"]>>;
  }

  getTeacher(id: string): Promise<TeacherModel> {
    return this.get(`${this.context.enviroment}/teachers/${id}.json`);
  }

  getStudentReservation(codigo: string): Promise<StudentReservationModel> {
    return this.get(
      `/${this.context.enviroment}/studentsReservations/${codigo}.json`
    );
  }

  async makeReservation(
    codigo: string,
    teacher_id: string,
    option_id: string,
    tutorial_reason?: string
  ) {
    const option = await this.getOptionById(option_id);
    if (option === null)
      throw new Error("option with id " + option_id + " returned null");
    const student = await this.context.dataSources.studentsAPI.getStudent(
      codigo
    );
    const { nombre, apellido_paterno, apellido_materno, nivel, grupo } =
      student;
    const teacherReservation: TeacherReservationModel = {
      codigo: student.codigo,
      nombre,
      apellido_paterno,
      apellido_materno,
      nivel,
      grupo,
      option_id: option.id,
      option_name: "why? option_name from makeReservation dataSource",
      workshop_id: option.workshop_id,
      workshop_name: option.workshop_name,
      tutorial_reason: tutorial_reason ? tutorial_reason : null,
    };
    this.post(
      `${this.context.enviroment}/teachers/${teacher_id}/raw_reservations/${option_id}.json`,
      teacherReservation
    );
    this.put(`${this.context.enviroment}/studentsReservations/${codigo}.json`, {
      option_id: option_id,
    });

    //enqueue
    const availableNumber = await this.get(
      `${this.context.enviroment}/availableOptions/${option_id}.json`
    );
    this.put(
      `${this.context.enviroment}/availableOptions/${option_id}.json`,
      availableNumber + 1
    );

    return { ...option, option_id: option.id };
  }

  deleteReservations(teacher_id: string, option_id: string) {
    return this.delete(
      `${this.context.enviroment}/teachers/${teacher_id}/raw_reservations/${option_id}.json`
    );
  }

  async resetReservations() {
    await this.delete(`${this.context.enviroment}/studentsReservations.json`);
    await this.delete(`${this.context.enviroment}/availableOptions.json`);
    return true;
  }

  getOptionById(id: string): Promise<OptionModel> {
    return this.get(`${this.context.enviroment}/options/${id}.json`);
  }

  async saveAttendance(attendance: AttendingStudent[]) {
    const date = new Date();
    const values = attendance.map((student) => {
      return [
        `=date(${date.getFullYear()},${date.getMonth() + 1},${date.getDate()})`,
        student.codigo,
        student.nombre,
        student.apellido_paterno,
        student.apellido_materno,
        student.nivel,
        student.grupo,
        student.workshop,
        student.teacher,
        student.attended,
      ];
    });
    const range = "Attendance!A1";

    try {
      await this.context.dataSources.workshopsSheetsAPI.append(values, range);
    } catch (e) {
      console.error(e);
    }
    return true;
  }

  setWorkshopLink(option_id: string, url: string) {
    this.put(`${this.context.enviroment}/options/${option_id}/url.json`, url);
    return true;
  }

  async getMaxStudentReservations() {
    const result = await this.get(`${this.context.enviroment}/system.json`);
    return result.max_reservations;
  }

  async getReservationLimit(): Promise<number> {
    return this.get("/reservationLimit.json");
  }
}

export { WorkshopsAPI };
