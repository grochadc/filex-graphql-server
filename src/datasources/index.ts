import { FirebaseAPI } from "./myREST";
import { RegistroAPI } from "./RegistroAPI";
import { ExamAPI } from "./ExamAPI";
import { WorkshopsAPI } from "./WorkshopsAPI";
import { StudentsAPI } from "./StudentsAPI";

export { RegistroAPI, FirebaseAPI, ExamAPI, WorkshopsAPI, StudentsAPI };
