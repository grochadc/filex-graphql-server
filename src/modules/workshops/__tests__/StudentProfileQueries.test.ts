import { expect, jest, test } from "@jest/globals";
import { StudentsAPI } from "../../../datasources/StudentsAPI";
jest.mock("../../../datasources/StudentsAPI");
import { WorkshopsAPI } from "../../../datasources/WorkshopsAPI";
jest.mock("../../../datasources/WorkshopsAPI");

import testServer from "../../../testUtils/testServer";
import { gql } from "apollo-server";
import getStudentReservationMock from "../../../datasources/WorkshopsAPI/mocks/getStudentReservationPrismaMock";
import getStudentMock from "../../../datasources/WorkshopsAPI/mocks/studentPrismaMock";
import prisma from "../../../datasources/testutils/prisma/client";

const studentsAPI = new StudentsAPI({} as any, prisma);
const workshopsAPI = new WorkshopsAPI(prisma);
let server: ReturnType<typeof testServer>;
beforeEach(() => {
  //@ts-ignore
  studentsAPI.getStudent.mockResolvedValue(getStudentMock);
  //@ts-ignore
  workshopsAPI.getStudentReservation.mockResolvedValue(
    getStudentReservationMock
  );
  //@ts-ignore
  workshopsAPI.getReservationCount.mockResolvedValue(0);
  //@ts-ignore
  workshopsAPI.getReservationLimit.mockResolvedValue(5);

  server = testServer(() => {
    return {
      studentsAPI,
      workshopsAPI,
    };
  });
});

describe("[Unit Tests] Gets a student", () => {
  it("full info", async () => {
    const result = await server.query({
      query: gql`
        query getStudent($codigo: ID!) {
          student(codigo: $codigo) {
            id
            codigo
            nombre
            apellido_paterno
            apellido_materno
            ciclo
            email
            telefono
            nivel
          }
        }
      `,
      variables: { codigo: "1234567890" },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data.student.id).toBe(3);
    expect(result.data).toMatchSnapshot();
  });

  it("reservation", async () => {
    const result = await server.query({
      query: gql`
        query getStudentReservation($codigo: ID!) {
          student(codigo: $codigo) {
            reservation {
              option {
                workshop {
                  name
                }
                day
                time
                url
                teacher {
                  nombre
                }
              }
            }
          }
        }
      `,
      variables: { codigo: "1234567890" },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data).toMatchSnapshot();
  });

  it("reservationCount and reservationLimit", async () => {
    const result = await server.query({
      query: gql`
        query countStudentReservations($codigo: ID!) {
          student(codigo: $codigo) {
            reservationCount
            reservationLimit
          }
        }
      `,
      variables: { codigo: "1234567890" },
    });

    expect(result.errors).toBeUndefined();
    expect(result.data).toMatchSnapshot();
  });
});

test("[Integration Test] gets a student profile", async () => {
  const result = await server.query({
    query: gql`
      query getStudentProfile($code: ID!) {
        student(codigo: $code) {
          id
          codigo
          nombre
          apellido_paterno
          apellido_materno
          email
          telefono
          nivel
          reservation {
            option {
              workshop {
                name
              }
              day
              time
              url
              teacher {
                nombre
              }
            }
          }
          reservationCount
          reservationLimit
        }
      }
    `,
    variables: { code: "1234567890" },
  });

  expect(WorkshopsAPI).toHaveBeenCalled();
  expect(workshopsAPI.getStudentReservation).toBeCalled();
  expect(workshopsAPI.getReservationCount).toHaveBeenCalled();
  expect(workshopsAPI.getReservationLimit).toHaveBeenCalled();
  expect(result.errors).toBeUndefined();
  expect(result.data).toMatchSnapshot();
});
