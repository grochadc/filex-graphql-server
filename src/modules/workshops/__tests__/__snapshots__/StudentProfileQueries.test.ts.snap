// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Integration Test] gets a student profile 1`] = `
{
  "student": {
    "apellido_materno": "Ocasio",
    "apellido_paterno": "Martinez",
    "codigo": "1234567890",
    "email": "bad@bunny.pr",
    "id": "st_3",
    "nivel": 4,
    "nombre": "Benito Antonio",
    "reservation": {
      "option": {
        "day": "Lunes",
        "teacher": {
          "nombre": "Gonzalo Rocha",
        },
        "time": "13:00",
        "url": null,
        "workshop": {
          "name": "Conversation",
        },
      },
    },
    "reservationCount": 0,
    "reservationLimit": 5,
    "telefono": "0001234567",
  },
}
`;

exports[`[Unit Tests] Gets a student full info 1`] = `
{
  "student": {
    "apellido_materno": "Ocasio",
    "apellido_paterno": "Martinez",
    "ciclo": "2022A",
    "codigo": "1234567890",
    "email": "bad@bunny.pr",
    "id": "st_3",
    "nivel": 4,
    "nombre": "Benito Antonio",
    "telefono": "0001234567",
  },
}
`;

exports[`[Unit Tests] Gets a student reservation 1`] = `
{
  "student": {
    "reservation": {
      "option": {
        "day": "Lunes",
        "teacher": {
          "nombre": "Gonzalo Rocha",
        },
        "time": "13:00",
        "url": null,
        "workshop": {
          "name": "Conversation",
        },
      },
    },
  },
}
`;

exports[`[Unit Tests] Gets a student reservationCount and reservationLimit 1`] = `
{
  "student": {
    "reservationCount": 0,
    "reservationLimit": 5,
  },
}
`;
